pkg/nimble/netif: allow to apply and enforce random and unique connection intervals

haukepetersen opened the pull request:
### Contribution description
As we were able to show in our research, using the same connection interval for multiple connections of a node can decrease network performance significantly. To prevent this, one solution is to use random connection intervals chosen from a configured range. This is already in the RIOT code base specifically for the `nimble_statconn` module. This PR generalizes this approach to be applicable for all modules using `nimble_netif`.

On top, this PR adds the option to enforce different connection intervals for each of a nodes connection. If a nodes opens a new connection as master, it will generate random intervals and check them against the interval of each existing connections, until it finds a connection interval that is spaced at least `NIMBLE_NETIF_CONN_ITVL_SPACING_MS` ticks apart from the ones of the existing connections. If a node is accepting a new connection in the slave role, it will also check the connection interval of the newly created connection against the intervals of all of its existing connections. If the interval of the new connection is not unique w.r.t. to the spacing contraints, the newly created connection is closed again.

As it improves the network behavior quite a bit, this enforcing feature (`NIMBLE_NETIF_FORCE_CONN_ITVL_SPACING`)  is enabled by default and is used anytime a range for minimum and maximum connection intervals is configured.

### Testing procedure
a) verify using vanilla `gnrc_networking` using `nimble_netif` that everything is working as before

b) run `gnrc_networking` using `nimble_autoconn` and specifying a connection interval range, e.g.
`CFLAGS="-DNIMBLE_AUTOCONN_CONN_ITVL_MIN_MS=50 -DNIMBLE_AUTOCONN_CONN_ITVL_MAX_MS=70" make ...`
on at least two BLE nodes in radio range. The nodes should automatically open a connection. If you close the connection using `ble close X`  the connection should be re-opened. Everytime this happens, the connection should have a random connection interval from the given range applied.

c) do the same using `nimble_statconn` -> when manually closing a connection it should be re-established with a random connection interval from the specified range.

### Issues/PRs references
none

haukepetersen commented:
added some doxygen improvement and a fix when using `NULL` as connection parameters in `nimble_netif_connect()`.

haukepetersen commented:
rebased

fjmolinas commented on pkg/nimble/netif/include/nimble_netif_conn.h:
```diff
@@ -175,6 +175,57 @@ int nimble_netif_conn_start_adv(void);
+ *          with the connection interval of another BLE connection
+ * @return  false if the connection interval of the given connection is valid
+ */
+int nimble_netif_conn_itvl_invalid(int handle);
```
```suggestion
bool nimble_netif_conn_itvl_invalid(int handle);
```

why not bool?

fjmolinas commented on pkg/nimble/netif/nimble_netif_conn.c:
```diff
@@ -220,3 +221,73 @@ unsigned nimble_netif_conn_count(uint16_t filter)
+{
+    assert(min <= max);
+
+    uint16_t start = random_uint32_range(min, max);
+    if (IS_ACTIVE(NIMBLE_NETIF_FORCE_CONN_ITVL_SPACING)) {
+        for (uint16_t itvl = start;
+             itvl <= max;
+             itvl += NIMBLE_NETIF_CONN_ITVL_SPACING_MS) {
+            if (!nimble_netif_conn_itvl_used(itvl, NIMBLE_NETIF_CONN_INVALID)) {
+                return itvl;
+            }
+        }
+        for (uint16_t itvl = start - NIMBLE_NETIF_CONN_ITVL_SPACING_MS;
+             itvl >= min;
+             itvl -= NIMBLE_NETIF_CONN_ITVL_SPACING_MS) {
+            if (!nimble_netif_conn_itvl_used(itvl, NIMBLE_NETIF_CONN_INVALID)) {
+                return itvl;
+            }
+        }
+        return 0;
+    }
```
So if I understand this correctly this is randomly picking a start point in the range, then search upwards for a valid interval, if it does not find it it searches downwards?

haukepetersen replied:
correct. Compared to choosing a random value in a loop, this has the huge advantage of a fully deterministic runtime, and also it will return an error if not suitable conn itvl is available anymore.

fjmolinas commented on pkg/nimble/netif/include/nimble_netif_conn.h:
```diff
@@ -175,6 +175,57 @@ int nimble_netif_conn_start_adv(void);
+int nimble_netif_conn_itvl_invalid(int handle);
+
+/**
+ * @brief   Generate a random connection interval from the given range
```
```suggestion
 * @brief   Generate a pseudorandom connection interval from the given range
```

haukepetersen replied:
fixed.

fjmolinas reviewed (commented):
Some comments, will test ASAP

fjmolinas commented:
> a) verify using vanilla gnrc_networking using nimble_netif that everything is working as before

```
2021-06-15 13:39:42,062 # NETOPT_RX_END_IRQ not implemented by driver
2021-06-15 13:39:42,063 # NETOPT_TX_END_IRQ not implemented by driver
2021-06-15 13:39:42,064 # main(): This is RIOT! (Version: 2021.07-devel-342-gf68ee1-pr-16372)
2021-06-15 13:39:42,065 # RIOT network stack example application
2021-06-15 13:39:42,065 # All up, running the shell now
> ifconfig
2021-06-15 13:39:50,118 # ifconfig
2021-06-15 13:39:50,121 # Iface  8  HWaddr: D8:2A:B8:E0:AB:50
2021-06-15 13:39:50,126 #           L2-PDU:1280  MTU:1280  HL:64  RTR
2021-06-15 13:39:50,128 #           6LO  IPHC
2021-06-15 13:39:50,131 #           Source address length: 6
2021-06-15 13:39:50,133 #           Link type: wireless
2021-06-15 13:39:50,139 #           inet6 addr: fe80::d82a:b8ff:fee0:ab50  scope: link  VAL
2021-06-15 13:39:50,142 #           inet6 group: ff02::2
2021-06-15 13:39:50,145 #           inet6 group: ff02::1
2021-06-15 13:39:50,148 #           inet6 group: ff02::1:ffe0:ab50
2021-06-15 13:39:50,149 #
2021-06-15 13:39:50,152 #           Statistics for Layer 2
2021-06-15 13:39:50,155 #             RX packets 0  bytes 0
2021-06-15 13:39:50,160 #             TX packets 0 (Multicast: 0)  bytes 9320
2021-06-15 13:39:50,163 #             TX succeeded 0 errors 0
2021-06-15 13:39:50,165 #           Statistics for IPv6
2021-06-15 13:39:50,169 #             RX packets 0  bytes 0
2021-06-15 13:39:50,173 #             TX packets 2 (Multicast: 2)  bytes 112
2021-06-15 13:39:50,177 #             TX succeeded 2 errors 0
2021-06-15 13:39:50,177 #
> ble
2021-06-15 13:39:52,607 # ble
2021-06-15 13:39:52,613 # usage: ble [help|info|adv|scan|connect|close|update|chanmap]
> ble adv
2021-06-15 13:39:55,011 # ble adv
2021-06-15 13:39:55,015 # success: advertising this node as 'bleRIOT'
> 2021-06-15 13:40:08,238 # event: handle 0 -> CONNECTED as SLAVE (E4:DD:E0:8F:73:65)
ping6 fe80::e4dd:e0ff:fe8f:7365
2021-06-15 13:40:54,652 # ping6 fe80::e4dd:e0ff:fe8f:7365
2021-06-15 13:40:54,736 # 12 bytes from fe80::e4dd:e0ff:fe8f:7365%8: icmp_seq=0 ttl=64 time=77.189 ms
2021-06-15 13:40:55,736 # 12 bytes from fe80::e4dd:e0ff:fe8f:7365%8: icmp_seq=1 ttl=64 time=77.246 ms
2021-06-15 13:40:56,736 # 12 bytes from fe80::e4dd:e0ff:fe8f:7365%8: icmp_seq=2 ttl=64 time=76.779 ms
2021-06-15 13:40:56,736 #
2021-06-15 13:40:56,740 # --- fe80::e4dd:e0ff:fe8f:7365 PING statistics ---
2021-06-15 13:40:56,746 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-06-15 13:40:56,750 # round-trip min/avg/max = 76.779/77.071/77.246 ms
>
```

fjmolinas commented:
> b) run gnrc_networking using nimble_autoconn and specifying a connection interval range, e.g.
CFLAGS="-DNIMBLE_AUTOCONN_CONN_ITVL_MIN_MS=50 -DNIMBLE_AUTOCONN_CONN_ITVL_MAX_MS=70" make ...
on at least two BLE nodes in radio range. The nodes should automatically open a connection. If you close the connection using ble close X the connection should be re-opened. Everytime this happens, the connection should have a random connection interval from the given range applied.

```
> ble info
2021-06-15 13:43:41,546 # ble info
2021-06-15 13:43:41,551 # Own Address: E4:DD:E0:8F:73:65 -> [FE80::E4DD:E0FF:FE8F:7365]
2021-06-15 13:43:41,552 #  Free slots: 1/3
2021-06-15 13:43:41,554 # Advertising: yes
2021-06-15 13:43:41,555 # Connections: 1
2021-06-15 13:43:41,561 # [ 0] D8:2A:B8:E0:AB:50 [FE80::D82A:B8FF:FEE0:AB50] (M,56ms,2500ms,0)
2021-06-15 13:43:41,566 #      (role, conn itvl, superv. timeout, slave latency)
2021-06-15 13:43:41,567 # Slots:
2021-06-15 13:43:41,571 # [ 0] state: 0x0011 - GAP-master L2CAP-client
2021-06-15 13:43:41,574 # [ 1] state: 0x0100 - advertising
2021-06-15 13:43:41,576 # [ 2] state: 0x8000 - unused
2021-06-15 13:43:41,576 #
ble close
2021-06-15 13:44:00,140 # ble close
2021-06-15 13:44:00,143 # usage: ble close [help|list|<conn #>]
> ble close 0
2021-06-15 13:44:03,619 # ble close 0
2021-06-15 13:44:03,623 # success: connection tear down initiated
> ble info
2021-06-15 13:44:07,174 # ble info
2021-06-15 13:44:07,179 # Own Address: E4:DD:E0:8F:73:65 -> [FE80::E4DD:E0FF:FE8F:7365]
2021-06-15 13:44:07,181 #  Free slots: 3/3
2021-06-15 13:44:07,182 # Advertising: no
2021-06-15 13:44:07,184 # Connections: 0
2021-06-15 13:44:07,184 # Slots:
2021-06-15 13:44:07,186 # [ 0] state: 0x8000 - unused
2021-06-15 13:44:07,189 # [ 1] state: 0x8000 - unused
2021-06-15 13:44:07,191 # [ 2] state: 0x8000 - unused
2021-06-15 13:44:07,192 #
> ble info
2021-06-15 13:44:10,187 # ble info
2021-06-15 13:44:10,193 # Own Address: E4:DD:E0:8F:73:65 -> [FE80::E4DD:E0FF:FE8F:7365]
2021-06-15 13:44:10,193 #  Free slots: 1/3
2021-06-15 13:44:10,195 # Advertising: yes
2021-06-15 13:44:10,196 # Connections: 1
2021-06-15 13:44:10,202 # [ 0] D8:2A:B8:E0:AB:50 [FE80::D82A:B8FF:FEE0:AB50] (S,52ms,2500ms,0)
2021-06-15 13:44:10,208 #      (role, conn itvl, superv. timeout, slave latency)
2021-06-15 13:44:10,208 # Slots:
2021-06-15 13:44:10,211 # [ 0] state: 0x0022 - GAP-slave L2CAP-server
2021-06-15 13:44:10,215 # [ 1] state: 0x0100 - advertising
2021-06-15 13:44:10,217 # [ 2] state: 0x8000 - unused
ping6  fe80::d82a:b8ff:fee0:ab50
2021-06-15 13:44:33,538 # ping6  fe80::d82a:b8ff:fee0:ab50
2021-06-15 13:44:33,647 # 12 bytes from fe80::d82a:b8ff:fee0:ab50%8: icmp_seq=0 ttl=64 time=102.563 ms
2021-06-15 13:44:34,644 # 12 bytes from fe80::d82a:b8ff:fee0:ab50%8: icmp_seq=1 ttl=64 time=101.870 ms
2021-06-15 13:44:35,642 # 12 bytes from fe80::d82a:b8ff:fee0:ab50%8: icmp_seq=2 ttl=64 time=101.000 ms
2021-06-15 13:44:35,642 #
2021-06-15 13:44:35,646 # --- fe80::d82a:b8ff:fee0:ab50 PING statistics ---
2021-06-15 13:44:35,652 # 3 packets transmitted, 3 packets received, 0% packet loss
2021-06-15 13:44:35,656 # round-trip min/avg/max = 101.000/101.811/102.563 ms
```

fjmolinas commented:
> c) do the same using nimble_statconn -> when manually closing a connection it should be re-established with a random connection interval from the specified range.

Does this mean doing `statconn addm/adds` and the `ble close n`? Because doing that I always get the same interval.

haukepetersen commented on pkg/nimble/netif/include/nimble_netif.h:
```diff
@@ -102,6 +102,26 @@ extern "C" {
+ *          intervals, in multiples of 1.25ms
+ */
+#ifndef NIMBLE_NETIF_CONN_ITVL_SPACING_MS
+#define NIMBLE_NETIF_CONN_ITVL_SPACING_MS       2
```
True, the `_MS` suffix does not really apply here. I just saw that it would further simplify things if we simply merge `NIMBLE_NETIF_CONN_ITVL_SPACING_MS` and `NIMBLE_NETIF_FORCE_CONN_ITVL_SPACING` into a single variable. Will fix.

haukepetersen replied:
Also, the internal code of the NimBLE controller will convert this value to microseconds, hence a value o 2 here will lead to 2500 in the controller, circumventing the floating issue :-)

This 1.25ms factor is something that the Bluetooth standard defines, and this is also used in the HCI interface commands. So this int value is plainly propagated in the BLE Host.

haukepetersen commented:
Addressed comments:
- enhanced the connection interval spacing configuration: it is now configured using a single configuration value. Set to 0 to disable enforcing connection spacing...
- the default value for the connection spacing is now `0`. This has a simple reason: if a node is connection subordinate, it will check the connection interval spacing depending on this config option. But if the connection intervals in the application are defined with `itvl_min == itvl_max` (as for example in the vanilla `gnrc_networking`), the current code would allow only for a single connection...

> Does this mean doing statconn addm/adds and the ble close n? Because doing that I always get the same interval.

Yes, calling `close N` was the correct call. In the last commit I fixed a bug that was causing this interval being the same on every call: the `netif_connect` code does override the `itvl_min` and `itvl_max` values in the connection parameter struct with the random generated value before passing this to NimBLE. But `statconn` did use a static allocated params struct, that is reused for any (re)connection event. Now that the min/max values were overridden by `nimble_netif_connect`, this lead to the first generated random value to be applied for all subsequent connections. I fixed this by restoring the original values of min/max after doing the NimBLE API call in `nimble_netif_connect()`.

fjmolinas commented:
>c) do the same using nimble_statconn -> when manually closing a connection it should be re-established with a random connection interval from the specified range.

Statconn now works as expected!

```
statconn addm D8:2A:B8:E0:AB:50
2021-06-17 09:59:28,505 # statconn addm D8:2A:B8:E0:AB:50
2021-06-17 09:59:28,508 # success: connecting to peer as slave
> ble info
2021-06-17 09:59:33,391 # ble info
2021-06-17 09:59:33,397 # Own Address: E4:DD:E0:8F:73:65 -> [FE80::E4DD:E0FF:FE8F:7365]
2021-06-17 09:59:33,399 #  Free slots: 2/3
2021-06-17 09:59:33,400 # Advertising: no
2021-06-17 09:59:33,401 # Connections: 1
2021-06-17 09:59:33,407 # [ 0] D8:2A:B8:E0:AB:50 [FE80::D82A:B8FF:FEE0:AB50] (S,46ms,2500ms,0)
2021-06-17 09:59:33,412 #      (role, conn itvl, superv. timeout, slave latency)
2021-06-17 09:59:33,413 # Slots:
2021-06-17 09:59:33,417 # [ 0] state: 0x0022 - GAP-slave L2CAP-server
2021-06-17 09:59:33,419 # [ 1] state: 0x8000 - unused
2021-06-17 09:59:33,422 # [ 2] state: 0x8000 - unused
2021-06-17 09:59:33,422 #
> ble close 0
2021-06-17 09:59:40,826 # ble close 0
2021-06-17 09:59:40,829 # success: connection tear down initiated
> ble info
2021-06-17 09:59:47,112 # ble info
2021-06-17 09:59:47,118 # Own Address: E4:DD:E0:8F:73:65 -> [FE80::E4DD:E0FF:FE8F:7365]
2021-06-17 09:59:47,119 #  Free slots: 2/3
2021-06-17 09:59:47,121 # Advertising: no
2021-06-17 09:59:47,122 # Connections: 1
2021-06-17 09:59:47,128 # [ 0] D8:2A:B8:E0:AB:50 [FE80::D82A:B8FF:FEE0:AB50] (S,56ms,2500ms,0)
2021-06-17 09:59:47,133 #      (role, conn itvl, superv. timeout, slave latency)
2021-06-17 09:59:47,134 # Slots:
2021-06-17 09:59:47,137 # [ 0] state: 0x0022 - GAP-slave L2CAP-server
2021-06-17 09:59:47,140 # [ 1] state: 0x8000 - unused
2021-06-17 09:59:47,142 # [ 2] state: 0x8000 - unused
2021-06-17 09:59:47,143 #
```

fjmolinas commented:
please squash!

haukepetersen commented:
squashed.

fjmolinas reviewed (approved):
ACK!

haukepetersen commented:
Murdock noted, that the `random_uint32_range()` call in `nimble_netif_conn.c` could not be linked for some applications. Seems like I forgot to add that dependency into `pkg/nimble/Makefile.dep`... Fixed it and squashed afterwards.

haukepetersen commented:
Awesome to have this in, thanks for the review!
